### PR TITLE
Optimize schema sync task

### PR DESCRIPTION
**Pull request by yongman**
### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by yongman** [comment]
@iosmanthus PTAL~

File: dbms/src/TiDB/Schema/SchemaSyncService.cpp [hunk 1]
```diff
@@ -37,54 +37,92 @@ SchemaSyncService::SchemaSyncService(DB::Context & context_)
...
 {
+    // Add task for adding and removing keyspace sync schema tasks.
     handle = background_pool.addTask(
         [&, this] {
-            String stage;
             auto keyspaces = context.getTMTContext().getStorages().getAllKeyspaces();
-            KeyspaceID cur_ks = NullspaceID;
-            bool done_anything = false;
-            try
+            std::unique_lock<std::shared_mutex> lock(ks_map_mutex);
```
Comment (JaySon-Huang) [inline dbms/src/TiDB/Schema/SchemaSyncService.cpp]: Make this task function to be two sub-functions `SchemaSyncService::addKeyspaceGCTasks` and `SchemaSyncService::removeStaleKeyspaceTasks` seems better.
It is too long for a lambda function...

File: dbms/src/TiDB/Schema/SchemaSyncService.cpp [hunk 2]
```diff
@@ -37,54 +37,92 @@ SchemaSyncService::SchemaSyncService(DB::Context & context_)
...
-                LOG_ERROR(log, "[keyspace {}] {} failed by {} \n stack : {}", cur_ks, stage, e.displayText(), e.getStackTrace().toString());
-            }
-            catch (const Poco::Exception & e)
-            {
-                LOG_ERROR(log, "[keyspace {}] {} failed by {}", cur_ks, stage, e.displayText());
             }
-            catch (const std::exception & e)
+
+            // Remove stale sync schema task.
+            for (auto const & iter: ks_handle_map)
```
Comment (JaySon-Huang) [inline dbms/src/TiDB/Schema/SchemaSyncService.cpp]:
```suggestion
            for (auto const & [ks, task_handle]: ks_handle_map)
```

File: dbms/src/TiDB/Schema/SchemaSyncService.h [hunk 3]
```diff
@@ -62,6 +64,9 @@ class SchemaSyncService
     BackgroundProcessingPool & background_pool;
     BackgroundProcessingPool::TaskHandle handle;
 
+    mutable std::shared_mutex ks_map_mutex;
+    std::unordered_map<KeyspaceID, BackgroundProcessingPool::TaskHandle> ks_handle_map;
```
Comment (JaySon-Huang) [inline dbms/src/TiDB/Schema/SchemaSyncService.h:72]: add some comments about this

**Comment by yongman** [comment]
@JaySon-Huang PTAL~

**Review by iosmanthus — APPROVED** [review]

**Pull request merged**